pyomo/network/arc.py

```diff
@@ -0,0 +1,382 @@
+        else:
+            return IndexedArc.__new__(IndexedArc)
+
+    def __init__(self, *args, **kwds):
+        source = kwds.pop("source", kwds.pop("src", None))
+        destination = kwds.pop("destination", kwds.pop("dest", None))
+        ports = kwds.pop("ports", None)
+        self._directed = kwds.pop("directed", None)
+        self._rule = kwds.pop('rule', None)
+        kwds.setdefault("ctype", Arc)
+
+        super(Arc, self).__init__(*args, **kwds)
+
+        if source is destination is ports is None:
+            self._init_vals = None
+        else:
+            self._init_vals = dict(
+                source=source, destination=destination, ports=ports)
+
+    def construct(self, data=None):
+        """Initialize the Arc"""
+        if __debug__ and logger.isEnabledFor(logging.DEBUG):
+            logger.debug("Constructing Arc %s" % self.name)
+
+        if self._constructed:
+            return
+
+        timer = ConstructionTimer(self)
+        self._constructed = True
+
+        if self._rule is None and self._init_vals is None:
+            # No construction rule or values specified
+            return
+        elif self._rule is not None and self._init_vals is not None:
+            raise ValueError(
+                "Cannot specify rule along with source/destination/ports "
+                "keywords for arc '%s'" % self.name)
+
+        self_parent = self._parent()
+
+        if not self.is_indexed():
+            if self._rule is None:
+                tmp = self._init_vals
+            else:
+                try:
+                    tmp = self._rule(self_parent)
+                except Exception:
+                    err = sys.exc_info()[1]
+                    logger.error(
+                        "Rule failed when generating values for "
+                        "arc %s:\n%s: %s"
+                        % (self.name, type(err).__name__, err))
+                    raise
+            tmp = self._validate_init_vals(tmp)
+            self._setitem_when_not_present(None, tmp)
+        else:
+            if self._init_vals is not None:
+                raise IndexError(
+                    "Arc '%s': Cannot initialize multiple indices "
+                    "of an arc with single ports" % self.name)
+            for idx in self._index:
+                try:
+                    tmp = apply_indexed_rule(self, self._rule, self_parent, idx)
+                except Exception:
+                    err = sys.exc_info()[1]
+                    logger.error(
+                        "Rule failed when generating values for "
+                        "arc %s with index %s:\n%s: %s"
+                        % (self.name, str(idx), type(err).__name__, err))
+                    raise
+                tmp = self._validate_init_vals(tmp)
+                self._setitem_when_not_present(idx, tmp)
+        timer.report()
+
+    def deconstruct(self):
+        """Remove self from port lists"""
```

Comment:
Otherwise you would have to walk through every arc and collect all of the ones this port is a part of, which seemed like a waste to be, but @jsiirola what do you think?